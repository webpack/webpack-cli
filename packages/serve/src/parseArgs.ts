import logger from 'webpack-cli/lib/utils/logger';

type WebpackCLIType = {
    getCoreFlags: Function;
    argParser: Function;
};

type ArgsType = {
    devServerArgs: object;
    webpackArgs: object;
};

/**
 *
 * Parses raw dev server CLI args
 *
 * @param {Object} cli - webpack CLI object
 * @param {Object[]} devServerFlags - devServer flags
 * @param {String[]} args - unparsed devServer args processed from the CLI
 *
 * @returns {Object} parsed webpack args and dev server args objects
 */
export default function parseArgs(cli: WebpackCLIType, args: string[]): ArgsType {
    let devServerFlags: object[];
    try {
        devServerFlags = require('webpack-dev-server/bin/cli-flags').devServer;
    } catch (err) {
        throw new Error(`You need to install 'webpack-dev-server' for running 'webpack serve'.\n${err}`);
    }

    const core = cli.getCoreFlags();

    const parsedDevServerArgs = cli.argParser(devServerFlags, args, true);
    const devServerArgs = parsedDevServerArgs.opts;
    const parsedWebpackArgs = cli.argParser(core, parsedDevServerArgs.unknownArgs, true, process.title);
    const webpackArgs = parsedWebpackArgs.opts;

    // pass along the 'hot' argument to the dev server if it exists
    if (webpackArgs && webpackArgs.hot !== undefined) {
        devServerArgs['hot'] = webpackArgs.hot;
    }

    if (parsedWebpackArgs.unknownArgs.length > 0) {
        parsedWebpackArgs.unknownArgs
            .filter((e) => e)
            .forEach((unknown) => {
                logger.error(`Unknown argument: ${unknown}`);
            });
        process.exit(2);
    }

    return {
        devServerArgs,
        webpackArgs,
    };
}
