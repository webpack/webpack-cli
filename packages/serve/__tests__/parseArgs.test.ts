'use strict';

// TODO: update snapshots once we update to webpack-dev-server@4

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const errorMock: any = jest.fn();
jest.mock('webpack-cli/lib/utils/logger', () => {
    return {
        error: errorMock,
    };
});

import WebpackCLI from 'webpack-cli';
import parseArgs from '../src/parseArgs';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const processExitSpy: any = jest.spyOn(process, 'exit');
// eslint-disable-next-line @typescript-eslint/no-empty-function
processExitSpy.mockImplementation(() => {});

describe('parseArgs', () => {
    const cli = new WebpackCLI();

    beforeEach(() => {
        errorMock.mockClear();
        processExitSpy.mockClear();
    });

    it('parses webpack and dev server args', () => {
        const args = parseArgs(cli, ['--bonjour', '--target=node', '--port', '8080']);
        expect(args).toMatchSnapshot();
        expect(errorMock.mock.calls.length).toEqual(0);
        expect(processExitSpy.mock.calls.length).toEqual(0);
    });

    it('handles hot arg', () => {
        const args = parseArgs(cli, ['--hot']);
        expect(args).toMatchSnapshot();
        expect(errorMock.mock.calls.length).toEqual(0);
        expect(processExitSpy.mock.calls.length).toEqual(0);
    });

    it('handles unknown args', () => {
        parseArgs(cli, ['--unknown-arg', '--unknown-arg-2']);
        expect(errorMock.mock.calls).toMatchSnapshot();
        expect(processExitSpy.mock.calls.length).toEqual(1);
        expect(processExitSpy.mock.calls[0]).toEqual([2]);
    });
});
