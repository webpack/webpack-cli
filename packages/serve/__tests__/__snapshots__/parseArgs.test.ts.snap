// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseArgs handles hot arg 1`] = `
Object {
  "devServerArgs": Object {
    "clientLogLevel": "info",
    "hot": true,
    "inline": true,
    "liveReload": true,
    "serveIndex": true,
  },
  "webpackArgs": Object {
    "color": true,
    "config": Array [],
    "hot": true,
    "mode": "production",
  },
}
`;

exports[`parseArgs handles unknown args 1`] = `
Array [
  Array [
    "Unknown argument: --unknown-arg",
  ],
  Array [
    "Unknown argument: --unknown-arg-2",
  ],
]
`;

exports[`parseArgs parses webpack and dev server args 1`] = `
Object {
  "devServerArgs": Object {
    "bonjour": true,
    "clientLogLevel": "info",
    "inline": true,
    "liveReload": true,
    "port": 8080,
    "serveIndex": true,
  },
  "webpackArgs": Object {
    "color": true,
    "config": Array [],
    "mode": "production",
    "target": "node",
  },
}
`;
